### ISSUE TITLE
Evaluating N different trained agents on the same task/dataset at the same time

### ISSUE THREAD

**Issue by g-karthik** [facebookresearch/ParlAI#1543]
Hey guys, I have N *different* trained agents (with their own hyperparams) that are each capable of generating a response given dialog history and knowledge. I want to be able to run these agents in parallel on a shared validation/test dataset to generate N different agent responses given a constructed dialog history and knowledge from the aforementioned dataset, so I can take a look at the N responses in parallel and see how each of the N agents are doing.

My first question is: is it possible to accomplish the above with the framework? If not:

It seems to me that I'd need to add a new function in agents.py that internally calls `load_agent_module(opt)` N times, where the passed `opt` would be a deepcopy of an upper-level `opt` with the `model_file` argument changed to the path to one of the N trained agents' model file. This is similar to how you create `opt_singletask` in the `__init__` for `MultiWorld`. The upper-level `opt` would contain the paths to all N trained agent model files and the name of the task corresponding to the validation/test dataset. Consequently, I'd initialize a `MultiWorld` with a list of the N loaded agents and the upper-level `opt` -- the latter would get used to create the task agent/teacher since the task will be specified. Finally, I'd just parley over the `MultiWorld` to generate all the N different agent responses I'm looking for and store them.

Does this sound like the right approach to follow?

**Comment by g-karthik** [comment]
Oh wait, I don't think the above approach would work. If I pass the N agents to `__init__` in `MultiWorld`, it would simply create one world (since I have one task) and put all the N agents in that world. But the intention is not for my N agents to interact with each other, it's just to independently evaluate them on the same task. I guess I'd have to write my own variant of MultiWorld that enumerates the agents list passed to it instead of enumerating the comma-separated list of tasks (which happens to be 1, in my case) then?

**Comment by stephenroller** [comment]
You could create a sort of TeeAgent, whose observe() function which makes a copy of the observation dictionary and passes it along to each of the submodels. Then TeeAgent's act or batch would similarly fan out and collate the different responses.

You would need to manually build up separate opt dictionaries for each of the subagents, but this could be as easy as a list of model files and `running all_opts = [json.load(model_file + '.opt') for model_file in model_files]` Then, as you'd say, you'd make the N calls to create_agent:

https://github.com/facebookresearch/ParlAI/blob/1bfe2b1fa1f00b9c17b51d2223096118a0d36c89/parlai/core/agents.py#L519

**Comment by g-karthik** [comment]
Hey Stephen, what do you think about the approach I was describing though?

Also, since the evaluation I'm talking about is for all agents on the same task, I'm now thinking we could load all the opts, create N worlds corresponding to the agents, parley each world to get a model response, repeat until `world.epoch_done()` is True for all worlds.

**Comment by stephenroller** [comment]
What's your main goal? To evaluate all agents simultaneously? To do ensembling? Do you just want to compare N different agents?

Indeed MultiWorld assumes that all agents will observe all other agent's actions (like a chat room).

If you just want to monitor the N, I would just run eval_model/display_model N times and compare the outputs. A carefully constructed custom World that creates the N agents would also be reasonable.

**Comment by g-karthik** [comment]
My main goal is to create a list of examples of the following type: (dialog history, knowledge, model1_response, model2_response, ..., modelN_response), from some validation/test set.

**Comment by stephenroller** [comment]
I would build the TeeAgent like I suggested above
